File: leetcode/0020-valid-parentheses.cc
// https://leetcode.com/problems/valid-parentheses/
// Easy

#include<string>
#include<vector>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        vector<char> v;
        for (char &c: s) {
            if (c == '(') {
                v.push_back(')');
            } else if (c == '[') {
                v.push_back(']');
            } else if (c == '{') {
                v.push_back('}');
            } else if (v.size() == 0 || v.back() != c) {
                return false;
            } else {
                v.pop_back();
            }
        }
        return v.size() == 0;
    }
};

File: leetcode/python/0020-valid-parentheses.py
# https://leetcode.com/problems/valid-parentheses/
# Easy

class Solution:
    def isValid(self, s: str) -> bool:
        par = []
        for c in s:
            if c == '(':
                par.append(')')
            elif c == '[':
                par.append(']')
            elif c == '{':
                par.append('}')
            elif len(par) == 0 or par[-1] != c:
                return False
            else:
                par.pop()
        return len(par) == 0